solr/core/src/java/org/apache/solr/cloud/ZkController.java

```diff
@@ -2069,15 +2070,26 @@ private void waitForShardId(final CoreDescriptor cd) {
     if (log.isDebugEnabled()) {
       log.debug("waiting to find shard id in clusterstate for {}", cd.getName());
     }
+
+    Predicate<Replica> replicaPredicate = (Replica r) -> cd.getName().equals(r.getCoreName());
     try {
       DocCollection collection =
           zkStateReader.waitForState(
               cd.getCollectionName(),
               320,
               TimeUnit.SECONDS,
-              c -> c != null && c.getShardId(getNodeName(), cd.getName()) != null);
+              c ->
+                  c != null
+                      && c.getReplicasOnNode(getNodeName()) != null
```

Comment:
hmm; maybe 'c.getReplicasOnNode' should return an empty collection to avoid awkward code like this, which raises potential concerns about double-calling a method that perhaps might not trivially return something.   WDYT?